jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/PathResponseListener.java

```diff
@@ -0,0 +1,126 @@
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jetty.client.Response.Listener;
+import org.eclipse.jetty.http.HttpStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of {@link Response.ContentListener} that produces an {@link Path}
+ * that allows applications to save a file from a response {@link Response}
+ * like curl &lt;URL&gt; -o file.bin does.
+ * <p>
+ * Typical usage is:
+ * <pre>
+ *  httpClient.newRequest(host, port)
+ * .send(new PathResponseListener(Path.of("/tmp/file.bin"));
+ * 
+ *  var request = httpClient.newRequest(host, port);
+ *  CompletableFuture&gt;Path&gt; completable = PathResponseListener.write(request, Path.of("/tmp/file.bin"), rewriteExistingFile);
+ * </pre>
+ */
+public class PathResponseListener extends CompletableFuture<Path> implements Listener
+{
+    private static final Logger LOG = LoggerFactory.getLogger(InputStreamResponseListener.class);
+    
+    private final Path path;
+    private final FileChannel fileOut;
+    
+    public PathResponseListener(Path path, boolean overwrite) throws IOException
+    {           
+        this.path = path;
+        
+        // Throws the exception if file can't be overwritten 
+        // otherwise truncate it.
+        if (Files.exists(path) && !overwrite)
+        {
+            throw new FileAlreadyExistsException("File can't be overwritten");
+        }
+        
+        fileOut = FileChannel.open(this.path, StandardOpenOption.CREATE, StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING);
+    }
+    
+    @Override
+    public void onHeaders(Response response)
+    {
+        if (response.getStatus() != HttpStatus.OK_200)
+        {
+            this.cancel(true);
+            response.abort(new HttpResponseException(String.format("HTTP status code of response %d", response.getStatus()), response));
+        }
+    }
+    
+    @Override
+    public void onContent(Response response, ByteBuffer content)
+    {
+        try
+        {
+            var bytesWritten = fileOut.write(content);
+            if (LOG.isDebugEnabled())
+                LOG.debug("%d bytes written", bytesWritten);
+        }
+        catch (IOException e)
+        {
+            response.abort(e);
+        }
+    }
+
+    @Override
+    public void onComplete(Result result)
+    {
+        if (result.isFailed())
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Result failure", result.getFailure());
+            
+            completeExceptionally(result.getFailure());
+            return;
+        }
+        
+        this.complete(this.path);        
+    }
+    
+    public static CompletableFuture<Path> write(Request request, Path path, boolean overwrite)
+    {        
+        CompletableFuture<Path> l = null;
```

Comment:
Just make this variable be a `PathResponseListener`, so you don't need the cast.